### PR TITLE
[fixes #155] Improve MovablePackageCache debugging

### LINKED ISSUE #155: ENOENT: no such file or directory, lstat '/Users/spenner/l/pemberly-example-web_trunk/~'
It appears a `~` is getting into our build, I'll diagnose where exactly next, the failure displays itself as:

```sh
Build Error (WaitForTrees)

ENOENT: no such file or directory, lstat '/Users/spenner/l/pemberly-example-web_trunk/~'
```

<details>
  <summary>error.log</summary>

<pre>
=================================================================================

ENV Summary:

  TIME: Fri May 03 2019 07:09:04 GMT-0700 (Pacific Daylight Time)
  TITLE: ember
  ARGV:
  - /Users/spenner/.gradle/pemberly/node-v10.6.0-darwin-x64/bin/node
  - /Users/spenner/.config/node/default/bin/ember
  - b
  EXEC_PATH: /Users/spenner/.gradle/pemberly/node-v10.6.0-darwin-x64/bin/node
  TMPDIR: /var/folders/4r/whc65vwj1xggvvky3yy1cp9m000mw4/T
  SHELL: /usr/local/bin/fish
  PATH:
  - /Users/spenner/.gradle/pemberly/yarnpkg-1.15.2/bin
  - /Users/spenner/.gradle/pemberly/node-v10.6.0-darwin-x64/bin
  - /Users/spenner/l/pemberly-example-web_trunk/.gradle/pemberly
  - /Users/spenner/.cargo/bin
  - /Users/spenner/.fzf/bin
  - /Users/spenner/.yarn/bin
  - /Users/spenner/.local/bin
  - /Applications/Racket v7.1/bin/
  - /Users/spenner/.config/node/default/bin
  - /usr/local/bin
  - /usr/bin
  - /bin
  - /usr/sbin
  - /sbin
  - /usr/local/linkedin/bin
  - /Applications/VMware
  - Fusion.app/Contents/Public
  - /export/content/linkedin/bin
  - /export/content/granular/bin
  - /Applications/Wireshark.app/Contents/MacOS
  PLATFORM: darwin x64
  FREEMEM: 1014452224
  TOTALMEM: 17179869184
  UPTIME: 838790
  LOADAVG: 2.78857421875,2.7177734375,2.5830078125
  CPUS:
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  - Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz - 3100
  ENDIANNESS: LE
  VERSIONS:
  - ares: 1.14.0
  - cldr: 33.0
  - http_parser: 2.8.0
  - icu: 61.1
  - modules: 64
  - napi: 3
  - nghttp2: 1.32.0
  - node: 10.6.0
  - openssl: 1.1.0h
  - tz: 2018c
  - unicode: 10.0
  - uv: 1.21.0
  - v8: 6.7.288.46-node.13
  - zlib: 1.2.11

ERROR Summary:

  - broccoliBuilderErrorStack: Error: ENOENT: no such file or directory, lstat '/Users/spenner/l/pemberly-example-web_trunk/~'
  - code: [undefined]
  - codeFrame: ENOENT: no such file or directory, lstat '/Users/spenner/l/pemberly-example-web_trunk/~'
  - errorMessage: ENOENT: no such file or directory, lstat '/Users/spenner/l/pemberly-example-web_trunk/~'
        at WaitForTrees (@embroider/compat/addons)
-~- created here: -~-
    at new Plugin (/Users/spenner/src/embroider-build/embroider/node_modules/broccoli-plugin/index.js:7:31)
    at new WaitForTrees (/Users/spenner/src/embroider-build/embroider/packages/core/src/wait-for-trees.js:30:9)
    at new CompatAddons (/Users/spenner/src/embroider-build/embroider/packages/compat/src/compat-addons.js:42:21)
    at Object.defaultPipeline [as compatBuild] (/Users/spenner/src/embroider-build/embroider/packages/compat/src/default-pipeline.js:17:18)
    at module.exports (/Users/spenner/l/pemberly-example-web_trunk/ember-cli-build.js:64:41)
    at Builder.readBuildFile (/Users/spenner/l/pemberly-example-web_trunk/node_modules/ember-cli/lib/models/builder.js:52:14)
    at Builder.setupBroccoliBuilder (/Users/spenner/l/pemberly-example-web_trunk/node_modules/ember-cli/lib/models/builder.js:66:22)
    at new Builder (/Users/spenner/l/pemberly-example-web_trunk/node_modules/ember-cli/lib/models/builder.js:32:10)
    at BuildTask.run (/Users/spenner/l/pemberly-example-web_trunk/node_modules/ember-cli/lib/tasks/build.js:15:19)
    at Promise.resolve.then (/Users/spenner/l/pemberly-example-web_trunk/node_modules/ember-cli/lib/models/command.js:243:46)
    at tryCatcher (/Users/spenner/l/pemberly-example-web_trunk/node_modules/rsvp/dist/rsvp.js:323:19)
    at invokeCallback (/Users/spenner/l/pemberly-example-web_trunk/node_modules/rsvp/dist/rsvp.js:495:31)
    at /Users/spenner/l/pemberly-example-web_trunk/node_modules/rsvp/dist/rsvp.js:559:14
    at flush (/Users/spenner/l/pemberly-example-web_trunk/node_modules/rsvp/dist/rsvp.js:2402:5)
    at process._tickCallback (internal/process/next_tick.js:61:11)
    at Function.Module.runMain (internal/modules/cjs/loader.js:745:11)
    at startup (internal/bootstrap/node.js:236:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:560:3)
-~- (end) -~-
  - errorType: Build Error
  - location:
    - column: [undefined]
    - file: [undefined]
    - line: [undefined]
    - treeDir: [undefined]
  - message: ENOENT: no such file or directory, lstat '/Users/spenner/l/pemberly-example-web_trunk/~'
        at WaitForTrees (@embroider/compat/addons)
-~- created here: -~-
    at new Plugin (/Users/spenner/src/embroider-build/embroider/node_modules/broccoli-plugin/index.js:7:31)
    at new WaitForTrees (/Users/spenner/src/embroider-build/embroider/packages/core/src/wait-for-trees.js:30:9)
    at new CompatAddons (/Users/spenner/src/embroider-build/embroider/packages/compat/src/compat-addons.js:42:21)
    at Object.defaultPipeline [as compatBuild] (/Users/spenner/src/embroider-build/embroider/packages/compat/src/default-pipeline.js:17:18)
    at module.exports (/Users/spenner/l/pemberly-example-web_trunk/ember-cli-build.js:64:41)
    at Builder.readBuildFile (/Users/spenner/l/pemberly-example-web_trunk/node_modules/ember-cli/lib/models/builder.js:52:14)
    at Builder.setupBroccoliBuilder (/Users/spenner/l/pemberly-example-web_trunk/node_modules/ember-cli/lib/models/builder.js:66:22)
    at new Builder (/Users/spenner/l/pemberly-example-web_trunk/node_modules/ember-cli/lib/models/builder.js:32:10)
    at BuildTask.run (/Users/spenner/l/pemberly-example-web_trunk/node_modules/ember-cli/lib/tasks/build.js:15:19)
    at Promise.resolve.then (/Users/spenner/l/pemberly-example-web_trunk/node_modules/ember-cli/lib/models/command.js:243:46)
    at tryCatcher (/Users/spenner/l/pemberly-example-web_trunk/node_modules/rsvp/dist/rsvp.js:323:19)
    at invokeCallback (/Users/spenner/l/pemberly-example-web_trunk/node_modules/rsvp/dist/rsvp.js:495:31)
    at /Users/spenner/l/pemberly-example-web_trunk/node_modules/rsvp/dist/rsvp.js:559:14
    at flush (/Users/spenner/l/pemberly-example-web_trunk/node_modules/rsvp/dist/rsvp.js:2402:5)
    at process._tickCallback (internal/process/next_tick.js:61:11)
    at Function.Module.runMain (internal/modules/cjs/loader.js:745:11)
    at startup (internal/bootstrap/node.js:236:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:560:3)
-~- (end) -~-
  - name: BuildError
  - nodeAnnotation: @embroider/compat/addons
  - nodeName: WaitForTrees
  - originalErrorMessage: ENOENT: no such file or directory, lstat '/Users/spenner/l/pemberly-example-web_trunk/~'
  - stack: Error: ENOENT: no such file or directory, lstat '/Users/spenner/l/pemberly-example-web_trunk/~'

=================================================================================

</pre>
<details>

### PR DESCRIPTION
**Pull request by stefanpenner**
[fixes #155] Improve MovablePackageCache debugging

* remove async extra when not required. This improves stack-traces and debugging (and in general performs better)
* assert if the symlink’s target contains a ~

### CODE REVIEW

**Comment by buschtoens** [comment]
> and in general performs better

Saw this in my inbox and got curious. Do you have any sources for this?

**Comment by stefanpenner** [comment]
@buschtoens my own benchmarks, and analysis of the posix implementation of fs access in libuv

Interestingly though, win32 libuv stuff which utilizing infrastructure provided by windows, seemed to perform slightly faster.

In general though, if your use-case is OK with blocking the event loop, doing sequential sync disk IO is a good option, if your use-case IS NOT OK with blocking the even loop (like you handling many concurrent API requests, and you want to maximize concurrency over throughput) then async disk IO is your best bet.

File: packages/compat/src/moved-package-cache.ts [hunk 1]
```diff
@@ -1,9 +1,16 @@
...
-import { ensureSymlinkSync, readdir, readlink, realpath, lstat } from 'fs-extra';
+import { ensureSymlinkSync, readdirSync, readlinkSync, realpathSync, lstatSync } from 'fs-extra';
 import { Memoize } from 'typescript-memoize';
 import { PackageCache, Package, getOrCreate } from '@embroider/core';
 import { MacrosConfig } from '@embroider/macros';
 
+function assertNoTildeExpansion(source: string, target: string) {
+  if (target.includes('~')) {
+    throw new Error(
+      `The symbolic link: ${source}'s target: ${target} contained a bash expansion '~' which is not supported.`
```
Comment (stefanpenner) [inline packages/compat/src/moved-package-cache.ts]: @ef4 thoughts on the wording?

Comment (ef4) [reply packages/compat/src/moved-package-cache.ts]: Looks good.

**Pull request merged**